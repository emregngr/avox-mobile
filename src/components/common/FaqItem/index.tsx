import { MaterialCommunityIcons } from '@expo/vector-icons'
import { useMemo } from 'react'
import { TouchableOpacity, View } from 'react-native'
import Collapsible from 'react-native-collapsible'

import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import type { FaqItemType } from '@/types/common/faq'

interface FaqItemProps {
  index: number
  isExpanded: boolean
  item: FaqItemType
  toggleExpanded: (index: number) => void
}

export const FaqItem = ({ index, isExpanded, item, toggleExpanded }: FaqItemProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const { description, title } = item
  return (
    <View className="mb-2">
      <TouchableOpacity
        activeOpacity={0.7}
        className="p-4 rounded-xl overflow-hidden bg-background-secondary"
        hitSlop={10}
        onPress={() => toggleExpanded(index)}
        testID={`faq-${item?.id}`}
      >
        <View className="flex-row justify-between items-center">
          <View className="flex-1 pr-4">
            <ThemedText color="text-100" type="body1">
              {title}
            </ThemedText>
          </View>
          <MaterialCommunityIcons
            color={colors?.onPrimary50}
            name={isExpanded ? 'chevron-up' : 'chevron-down'}
            size={24}
          />
        </View>
      </TouchableOpacity>

      <Collapsible collapsed={!isExpanded}>
        <View className="mt-2 p-4 rounded-xl overflow-hidden bg-background-tertiary">
          <ThemedText color="text-90" type="body2">
            {description}
          </ThemedText>
        </View>
      </Collapsible>
    </View>
  )
}
