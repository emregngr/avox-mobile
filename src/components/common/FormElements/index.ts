export * from './CheckboxField'
export * from './TextInputField'
