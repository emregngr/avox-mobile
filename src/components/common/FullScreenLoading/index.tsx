import React from 'react'
import { ActivityIndicator, View } from 'react-native'

export const FullScreenLoading = () => (
  <View
    className="flex-1 items-center justify-center bg-background-primary"
    testID="full-screen-loading"
  >
    <ActivityIndicator className="color-text-100" size="large" testID="loading-indicator" />
  </View>
)
