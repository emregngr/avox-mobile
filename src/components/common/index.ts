export * from './FaqItem'
export * from './FormElements'
export * from './FullScreenLoading'
export * from './Header'
export * from './OnboardingItem'
export * from './ProfileItem'
export * from './RenderDetailTabBar'
export * from './RenderTabBar'
export * from './SafeLayout'
export * from './SearchInput'
export * from './ThemedButton'
export * from './ThemedButtonText'
export * from './ThemedGradientButton'
export * from './ThemedText'
export * from './ToastConfig'
