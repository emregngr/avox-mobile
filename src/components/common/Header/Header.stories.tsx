import type { Meta, StoryObj } from '@storybook/react-native'
import React, { useState } from 'react'
import { Alert, TouchableOpacity, View } from 'react-native'

import { Header } from '@/components/common/Header'
import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import { MaterialCommunityIcons } from '@expo/vector-icons'

const meta: Meta<typeof Header> = {
  title: 'Components/Header',
  component: Header,
  parameters: {
    docs: {
      description: {
        component:
          'Header component with back button, title, and customizable right actions including glass effect styling.',
      },
    },
  },
  decorators: [
    Story => (
      <View className="flex-1 bg-background-primary pt-[20px]">
        <Story />
        <View className="flex-1 bg-background-secondary m-4 rounded-xl justify-center items-center">
          <ThemedText type="h3" color="text-100">
            Content Area
          </ThemedText>
        </View>
      </View>
    ),
  ],
  argTypes: {
    title: {
      control: { type: 'text' },
      description: 'Header title text',
    },
    backIcon: {
      control: { type: 'boolean' },
      description: 'Show back button',
    },
    rightButtonLabel: {
      control: { type: 'text' },
      description: 'Right button label text',
    },
    hapticFeedback: {
      control: { type: 'boolean' },
      description: 'Enable haptic feedback on interactions',
    },
    isFavorite: {
      control: { type: 'boolean' },
      description: 'Favorite state for haptic feedback intensity',
    },
    containerClassName: {
      control: { type: 'text' },
      description: 'Additional CSS classes for container',
    },
    titleClassName: {
      control: { type: 'text' },
      description: 'Additional CSS classes for title',
    },
    testID: {
      control: { type: 'text' },
      description: 'Test identifier',
    },
  },
}

export default meta

type Story = StoryObj<typeof Header>

export const Default: Story = {
  args: {
    title: 'Page Title',
  },
  parameters: {
    docs: {
      description: {
        story: 'Basic header with title and back button.',
      },
    },
  },
}

export const NoBackButton: Story = {
  args: {
    title: 'Settings',
    backIcon: false,
  },
  parameters: {
    docs: {
      description: {
        story: 'Header without back button for top-level pages.',
      },
    },
  },
}

export const WithRightButton: Story = {
  args: {
    title: 'Edit Profile',
    rightButtonLabel: 'Save',
    rightButtonOnPress: () => Alert.alert('Save', 'Save button pressed'),
  },
  parameters: {
    docs: {
      description: {
        story: 'Header with text button on the right side.',
      },
    },
  },
}

export const WithRightIcon: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    return (
      <Header
        title="Article Title"
        rightIcon={
          <MaterialCommunityIcons color={colors?.onPrimary100} name={'heart-outline'} size={20} />
        }
        rightIconOnPress={() => Alert.alert('Favorite', 'Added to favorites')}
        hapticFeedback={true}
      />
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Header with icon button on the right side with haptic feedback.',
      },
    },
  },
}

export const WithShareIcon: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    return (
      <Header
        title="Shared Content"
        shareIcon={
          <MaterialCommunityIcons
            color={colors?.onPrimary100}
            name={'share-variant-outline'}
            size={20}
          />
        }
        rightIcon={
          <MaterialCommunityIcons color={colors?.onPrimary100} name={'dots-horizontal'} size={20} />
        }
        shareIconOnPress={() => Alert.alert('Share', 'Share dialog opened')}
        rightIconOnPress={() => Alert.alert('More', 'More options menu')}
      />
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Header with both share and more icons.',
      },
    },
  },
}

export const LongTitle: Story = {
  args: {
    title:
      'This is a very long title that should be truncated properly when it exceeds the available space',
    rightButtonLabel: 'Done',
    rightButtonOnPress: () => Alert.alert('Done', 'Done button pressed'),
  },
  parameters: {
    docs: {
      description: {
        story: 'Header with long title that gets truncated.',
      },
    },
  },
}

export const NoTitle: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    return (
      <Header
        rightIcon={
          <MaterialCommunityIcons color={colors?.onPrimary100} name={'heart-outline'} size={20} />
        }
        rightIconOnPress={() => Alert.alert('Heart', 'Heart icon pressed')}
      />
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Header without title, showing only navigation elements.',
      },
    },
  },
}

export const AllFeatures: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    return (
      <Header
        title="Full Featured Header"
        shareIcon={
          <MaterialCommunityIcons
            color={colors?.onPrimary100}
            name={'share-variant-outline'}
            size={20}
          />
        }
        rightIcon={
          <MaterialCommunityIcons color={colors?.onPrimary100} name={'heart-outline'} size={20} />
        }
        backIconOnPress={() => Alert.alert('Back', 'Navigation back')}
        shareIconOnPress={() => Alert.alert('Share', 'Share content')}
        rightIconOnPress={() => Alert.alert('Favorite', 'Toggle favorite')}
        hapticFeedback={true}
        isFavorite={true}
      />
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Header with all available features enabled.',
      },
    },
  },
}

export const InteractiveStates: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    const [isFavorite, setIsFavorite] = useState<boolean>(false)
    const [title, setTitle] = useState<string>('Interactive Header')

    return (
      <Header
        title={title}
        backIconOnPress={() => Alert.alert('Back', 'Going back...')}
        rightIcon={
          <MaterialCommunityIcons
            color={isFavorite ? colors?.tertiary100 : colors?.onPrimary100}
            name={isFavorite ? 'heart' : 'heart-outline'}
            size={20}
          />
        }
        rightIconOnPress={() => {
          setIsFavorite(!isFavorite)
          setTitle(isFavorite ? 'Removed from favorites' : 'Added to favorites')
          setTimeout(() => setTitle('Interactive Header'), 2000)
        }}
        hapticFeedback={true}
        isFavorite={isFavorite}
      />
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Interactive header with state changes and feedback.',
      },
    },
  },
}

export const LayoutVariations: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    return (
      <View className="gap-4">
        <View>
          <ThemedText type="body4" color="text-70" className="mb-2">
            Back + Title
          </ThemedText>
          <Header title="Back with Title" />
        </View>
        <View>
          <ThemedText type="body4" color="text-70" className="mb-2">
            Title + Right Button
          </ThemedText>
          <Header
            title="Title with Button"
            backIcon={false}
            rightButtonLabel="Action"
            rightButtonOnPress={() => Alert.alert('Action', 'Button pressed')}
          />
        </View>
        <View>
          <ThemedText type="body4" color="text-70" className="mb-2">
            Title + Multiple Icons
          </ThemedText>
          <Header
            title="Multiple Icons"
            shareIcon={
              <MaterialCommunityIcons
                color={colors?.onPrimary100}
                name={'share-variant-outline'}
                size={20}
              />
            }
            shareIconOnPress={() => Alert.alert('Share', 'Share pressed')}
            rightIcon={
              <MaterialCommunityIcons
                color={colors?.onPrimary100}
                name={'dots-horizontal'}
                size={20}
              />
            }
            rightIconOnPress={() => Alert.alert('More', 'More pressed')}
          />
        </View>
        <View>
          <ThemedText type="body4" color="text-70" className="mb-2">
            Icons Only
          </ThemedText>
          <Header
            shareIcon={
              <MaterialCommunityIcons
                color={colors?.onPrimary100}
                name={'share-variant-outline'}
                size={20}
              />
            }
            shareIconOnPress={() => Alert.alert('Share', 'Share pressed')}
            rightIcon={
              <MaterialCommunityIcons
                color={colors?.onPrimary100}
                name={'heart-outline'}
                size={20}
              />
            }
            rightIconOnPress={() => Alert.alert('Heart', 'Heart pressed')}
          />
        </View>
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Different layout variations of the header component.',
      },
    },
  },
}

export const CustomStyling: Story = {
  args: {
    title: 'Custom Styled Header',
    containerClassName: 'bg-primary-100 rounded-xl mx-4 h-16 border-2 border-onPrimary-100',
    titleClassName: 'font-bold',
    rightButtonLabel: 'Custom',
    rightButtonOnPress: () => Alert.alert('Custom', 'Custom styled button'),
  },
  parameters: {
    docs: {
      description: {
        story: 'Header with custom styling and classes.',
      },
    },
  },
}

export const ErrorHandling: Story = {
  render: () => (
    <Header
      title="Error Test Header"
      rightButtonLabel="Error"
      rightButtonOnPress={() => {
        try {
          throw new Error('Test error')
        } catch (error) {
          Alert.alert('Error', 'Something went wrong!')
        }
      }}
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Header with error handling in button press.',
      },
    },
  },
}

export const Playground: Story = {
  argTypes: {
    rightIconType: {
      control: { type: 'select' },
      options: ['none', 'heart', 'share', 'more'],
      description: 'Select right icon type',
    },
    shareIconType: {
      control: { type: 'select' },
      options: ['none', 'heart', 'share', 'more'],
      description: 'Select share icon type',
    },
  } as any,
  render: args => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    const { rightIconType = 'heart', shareIconType = 'none', ...headerProps } = args as any

    return (
      <Header
        {...headerProps}
        rightIcon={
          rightIconType !== 'none' ? (
            <MaterialCommunityIcons color={colors?.onPrimary100} name={'heart-outline'} size={20} />
          ) : undefined
        }
        shareIcon={
          shareIconType !== 'none' ? (
            <MaterialCommunityIcons
              color={colors?.onPrimary100}
              name={'share-variant-outline'}
              size={20}
            />
          ) : undefined
        }
        rightIconOnPress={() => Alert.alert('Right Icon', `${rightIconType} icon pressed`)}
        shareIconOnPress={() => Alert.alert('Share Icon', `${shareIconType} icon pressed`)}
      />
    )
  },
  args: {
    title: 'Playground Header',
    backIcon: true,
    rightButtonLabel: '',
    hapticFeedback: true,
    isFavorite: false,
    testID: 'playground-header',
    rightIconType: 'heart',
    shareIconType: 'none',
  } as any,
  parameters: {
    docs: {
      description: {
        story: 'Interactive playground to test all props and configurations.',
      },
    },
  },
}

export const IconPlayground: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    const [selectedIcon, setSelectedIcon] = useState<
      'heart-outline' | 'share-variant-outline' | 'dots-horizontal'
    >('heart-outline')
    const [showShareIcon, setShowShareIcon] = useState<boolean>(false)

    const iconButtons = [
      { type: 'heart-outline' as const, label: 'Heart' },
      { type: 'share-variant-outline' as const, label: 'Share' },
      { type: 'dots-horizontal' as const, label: 'More' },
    ]

    return (
      <View>
        <Header
          title="Icon Playground"
          rightIcon={
            <MaterialCommunityIcons color={colors?.onPrimary100} name={selectedIcon} size={20} />
          }
          shareIcon={
            showShareIcon ? (
              <MaterialCommunityIcons
                color={colors?.onPrimary100}
                name={'share-variant-outline'}
                size={20}
              />
            ) : undefined
          }
          rightIconOnPress={() => Alert.alert('Right Icon', `${selectedIcon} pressed`)}
          shareIconOnPress={() => Alert.alert('Share', 'Share pressed')}
          hapticFeedback={true}
        />

        <View className="p-4 gap-3">
          <ThemedText type="body4" color="text-70">
            Right Icon Type:
          </ThemedText>
          <View className="flex-row gap-2">
            {iconButtons.map(icon => (
              <TouchableOpacity
                key={icon.type}
                activeOpacity={0.7}
                className="p-2 rounded min-w-[60px] items-center"
                style={{
                  backgroundColor:
                    selectedIcon === icon.type ? colors?.primary100 : colors?.secondary100,
                }}
                onPress={() => setSelectedIcon(icon.type)}
              >
                <ThemedText type="body4" color="text-100">
                  {icon.label}
                </ThemedText>
              </TouchableOpacity>
            ))}
          </View>

          <TouchableOpacity
            activeOpacity={0.7}
            className="p-2 rounded items-center mt-2"
            style={{ backgroundColor: showShareIcon ? colors?.primary100 : colors?.secondary100 }}
            onPress={() => setShowShareIcon(!showShareIcon)}
          >
            <ThemedText type="body4" color="text-100">
              {showShareIcon ? 'Hide' : 'Show'} Share Icon
            </ThemedText>
          </TouchableOpacity>
        </View>
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Interactive playground to test different icon combinations.',
      },
    },
  },
}

export const ButtonPlayground: Story = {
  render: () => {
    const { selectedTheme } = useThemeStore()
    const colors = themeColors?.[selectedTheme]

    const [buttonLabel, setButtonLabel] = useState<string>('Action')
    const [showButton, setShowButton] = useState<boolean>(true)

    const buttonLabels = ['Save', 'Done', 'Edit', 'More', 'Action']

    return (
      <View>
        <Header
          title="Button Playground"
          rightButtonLabel={showButton ? buttonLabel : ''}
          rightButtonOnPress={() => Alert.alert('Button', `${buttonLabel} pressed`)}
          hapticFeedback={true}
        />

        <View className="p-4 gap-3">
          <ThemedText type="body4" color="text-70">
            Button Labels:
          </ThemedText>
          <View className="flex-row flex-wrap gap-2">
            {buttonLabels.map(label => (
              <TouchableOpacity
                key={label}
                activeOpacity={0.7}
                className="p-2 rounded min-w-[50px] items-center"
                style={{
                  backgroundColor:
                    buttonLabel === label ? colors?.primary100 : colors?.secondary100,
                }}
                onPress={() => setButtonLabel(label)}
              >
                <ThemedText type="body4" color="text-100">
                  {label}
                </ThemedText>
              </TouchableOpacity>
            ))}
          </View>

          <TouchableOpacity
            activeOpacity={0.7}
            className="p-2 rounded items-center mt-2"
            style={{ backgroundColor: showButton ? colors?.primary100 : colors?.secondary100 }}
            onPress={() => setShowButton(!showButton)}
          >
            <ThemedText type="body4" color="text-100">
              {showButton ? 'Hide' : 'Show'} Button
            </ThemedText>
          </TouchableOpacity>
        </View>
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Interactive playground to test different button configurations.',
      },
    },
  },
}
