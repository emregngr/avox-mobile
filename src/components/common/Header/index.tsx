import { GlassView } from 'expo-glass-effect'
import * as Haptics from 'expo-haptics'
import { type ReactNode, useMemo } from 'react'
import type { ViewStyle } from 'react-native'
import { TouchableOpacity, View } from 'react-native'

import Back from '@/assets/icons/back.svg'
import { ThemedButtonText } from '@/components/common/ThemedButtonText'
import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import { cn } from '@/utils/common/cn'

type HeaderProps = {
  backIcon?: boolean
  backIconOnPress?: () => void
  containerClassName?: string
  hapticFeedback?: boolean
  isFavorite?: boolean
  rightButtonLabel?: string
  rightButtonOnPress?: () => void
  rightIcon?: ReactNode
  rightIconOnPress?: () => void
  shareIcon?: ReactNode
  shareIconOnPress?: () => void
  style?: ViewStyle
  testID?: string
  title?: string | string[]
  titleClassName?: string
}

export const Header = ({
  backIcon = true,
  backIconOnPress,
  containerClassName,
  hapticFeedback = false,
  isFavorite = false,
  rightButtonLabel,
  rightButtonOnPress,
  rightIcon,
  rightIconOnPress,
  shareIcon,
  shareIconOnPress,
  style,
  testID,
  title,
  titleClassName,
}: HeaderProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const STATIC_STYLES = {
    backIcon: {
      left: 16,
      position: 'absolute' as const,
      borderRadius: 20,
      padding: 8,
    },
    shareIcon: {
      right: 64,
      position: 'absolute' as const,
      borderRadius: 20,
      padding: 8,
    },
    rightIcon: {
      right: 16,
      position: 'absolute' as const,
      borderRadius: 20,
      padding: 8,
    },
  }

  const handleRightIconOnPress = () => {
    if (hapticFeedback) {
      if (isFavorite) {
        Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light)
      } else {
        Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Medium)
      }
    }

    rightIconOnPress?.()
  }

  return (
    <View
      className={cn('h-11 justify-center items-center', containerClassName)}
      style={style}
      testID={testID || 'header'}
    >
      {title ? (
        <ThemedText
          className={titleClassName ?? ''}
          color="text-100"
          ellipsizeMode="tail"
          numberOfLines={2}
          type="h3"
          center
        >
          {title}
        </ThemedText>
      ) : null}

      {backIcon ? (
        <GlassView
          style={STATIC_STYLES.backIcon}
          tintColor={colors?.background?.glass}
          glassEffectStyle="clear"
          isInteractive
        >
          <TouchableOpacity
            activeOpacity={0.7}
            hitSlop={20}
            onPress={backIconOnPress}
            testID="header-back-icon"
          >
            <Back color={colors?.onPrimary100} height={24} width={24} />
          </TouchableOpacity>
        </GlassView>
      ) : null}

      {shareIcon ? (
        <GlassView
          style={STATIC_STYLES.shareIcon}
          tintColor={colors?.background?.glass}
          glassEffectStyle="clear"
          isInteractive
        >
          <TouchableOpacity
            activeOpacity={0.7}
            hitSlop={10}
            onPress={shareIconOnPress}
            testID="header-share-icon"
          >
            {shareIcon}
          </TouchableOpacity>
        </GlassView>
      ) : null}

      {rightButtonLabel ? (
        <ThemedButtonText
          containerStyle="absolute right-4"
          label={rightButtonLabel}
          onPress={rightButtonOnPress as () => void}
          textColor="text-100"
          type="h4"
        />
      ) : rightIcon ? (
        <GlassView
          style={STATIC_STYLES.rightIcon}
          tintColor={colors?.background?.glass}
          glassEffectStyle="clear"
          isInteractive
        >
          <TouchableOpacity
            activeOpacity={0.7}
            hitSlop={10}
            onPress={handleRightIconOnPress}
            testID="header-right-icon"
          >
            {rightIcon}
          </TouchableOpacity>
        </GlassView>
      ) : null}
    </View>
  )
}
