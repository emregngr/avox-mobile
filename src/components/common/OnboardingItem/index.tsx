import { Image } from 'expo-image'
import { View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import type { OnBoardingType } from '@/types/common/onBoarding'

interface OnboardingItemProps {
  item: OnBoardingType
}

const STATIC_STYLES = {
  image: {
    height: '45%' as const,
    marginVertical: 20,
    width: '100%' as const,
  },
}

export const OnboardingItem = ({ item }: OnboardingItemProps) => {
  const { image, text, title } = item
  return (
    <View className="w-screen h-screen bg-background-primary">
      <Image
        cachePolicy="memory-disk"
        contentFit="contain"
        source={image}
        style={STATIC_STYLES.image}
        transition={0}
      />
      <View className="mx-4 z-10">
        <ThemedText
          className="mb-4"
          color="text-100"
          lineBreakMode="tail"
          numberOfLines={2}
          type="h1"
        >
          {title}
        </ThemedText>
        <ThemedText
          color="text-100" lineBreakMode="tail" numberOfLines={4}
          type="h3"
        >
          {text}
        </ThemedText>
      </View>
    </View>
  )
}
