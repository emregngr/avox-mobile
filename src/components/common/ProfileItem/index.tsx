import { MaterialCommunityIcons } from '@expo/vector-icons'
import type { ComponentType } from 'react'
import React, { useMemo } from 'react'
import { TouchableOpacity, View } from 'react-native'
import type { SvgProps } from 'react-native-svg'

import Right from '@/assets/icons/right.svg'
import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

type ProfileItemProps = {
  customLeftIcon?: ComponentType<SvgProps>
  danger?: boolean
  isLastItem?: boolean
  label: string
  leftIcon?: keyof typeof MaterialCommunityIcons.glyphMap
  onPress: () => void
  rightIcon?: boolean
  testID?: string
}

export const ProfileItem = ({
  customLeftIcon: CustomLeftIcon,
  danger,
  isLastItem = false,
  label,
  leftIcon,
  onPress,
  rightIcon = true,
  testID,
}: ProfileItemProps) => {
  const { selectedTheme } = useThemeStore()
  const colors = themeColors?.[selectedTheme]

  const iconColor = useMemo(
    () => (danger ? colors?.error : colors?.onPrimary100),
    [danger, colors?.error, colors?.onPrimary100],
  )

  const iconComponent = useMemo(() => {
    if (leftIcon) {
      return <MaterialCommunityIcons color={iconColor} name={leftIcon} size={24} />
    }
    if (CustomLeftIcon) {
      return <CustomLeftIcon color={iconColor} height={24} width={24} />
    }
    return null
  }, [leftIcon, CustomLeftIcon, iconColor])

  return (
    <>
      <TouchableOpacity
        activeOpacity={0.7}
        className="h-14 flex-row items-center justify-between px-4 bg-background-secondary"
        hitSlop={{ left: 20, right: 20 }}
        onPress={onPress}
        testID={testID || 'profile-item-touchable'}
      >
        <View className="flex-row items-center">
          {iconComponent}

          <ThemedText className="ml-4" color={danger ? 'error' : 'text-100'} type="body1">
            {label}
          </ThemedText>
        </View>

        {rightIcon ? (
          <Right
            color={iconColor} height={24} testID="right-icon"
            width={24}
          />
) : null}
      </TouchableOpacity>

      {!isLastItem ? (
        <View className="border-b border-background-quaternary ml-14" testID="separator" />
      ) : null}
    </>
  )
}
