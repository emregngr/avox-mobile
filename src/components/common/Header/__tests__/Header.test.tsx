import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'

import { Header } from '@/components/common/Header'
import useThemeStore from '@/store/theme'

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

const mockedBack = jest.fn()
const mockedRightButton = jest.fn()
const mockedRightIcon = jest.fn()
const mockedShareIcon = jest.fn()

beforeEach(() => {
  mockedUseThemeStore.mockReturnValue({
    selectedTheme: 'light',
  })
})

describe('Header Component', () => {
  it('renders the title when provided', () => {
    const { getByText } = render(<Header title="Test Title" />)
    expect(getByText('Test Title')).toBeTruthy()
  })

  it('does not render the title when not provided', () => {
    const { queryByText } = render(<Header />)
    expect(queryByText('Test Title')).toBeNull()
  })

  it('renders back icon by default and calls onPress when pressed', () => {
    const { getByTestId } = render(<Header backIconOnPress={mockedBack} title="Back Test" />)

    fireEvent.press(getByTestId('header-back-icon'))
    expect(mockedBack).toHaveBeenCalled()
  })

  it('does not render back icon when backIcon={false}', () => {
    const { queryByTestId } = render(<Header backIcon={false} />)
    expect(queryByTestId('header-back-icon')).toBeNull()
  })

  it('renders right button when label is provided and calls onPress', () => {
    const { getByText } = render(
      <Header rightButtonLabel="Save" rightButtonOnPress={mockedRightButton} />,
    )

    fireEvent.press(getByText('Save'))
    expect(mockedRightButton).toHaveBeenCalled()
  })

  it('renders right icon when provided and calls onPress', () => {
    const { getByTestId } = render(<Header rightIcon={<></>} rightIconOnPress={mockedRightIcon} />)

    fireEvent.press(getByTestId('header-right-icon'))
    expect(mockedRightIcon).toHaveBeenCalled()
  })

  it('renders share icon when provided and calls onPress', () => {
    const { getByTestId } = render(<Header shareIcon={<></>} shareIconOnPress={mockedShareIcon} />)

    fireEvent.press(getByTestId('header-share-icon'))
    expect(mockedShareIcon).toHaveBeenCalled()
  })
})

describe('Header Component Snapshot', () => {
  it('should render the Header Component successfully', () => {
    const { toJSON } = render(<Header title="Test Title" />)

    expect(toJSON()).toMatchSnapshot()
  })
})
