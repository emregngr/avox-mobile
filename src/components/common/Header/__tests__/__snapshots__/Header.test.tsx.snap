// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Header Component Snapshot should render the Header Component successfully 1`] = `
<View
  className="h-11 justify-center items-center"
  testID="header"
>
  <Text
    allowFontScaling={false}
    className="text-[18px] leading-[21px] font-inter-semibold text-center text-text-100"
    ellipsizeMode="tail"
    maxFontSizeMultiplier={1}
    numberOfLines={2}
  >
    Test Title
  </Text>
  <ViewManagerAdapter_ExpoGlassEffect
    glassEffectStyle="clear"
    isInteractive={true}
    style={
      {
        "borderRadius": 20,
        "left": 16,
        "padding": 8,
        "position": "absolute",
      }
    }
    tintColor="#00000015"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      hitSlop={20}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="header-back-icon"
    >
      <View
        color="#000000"
        height={24}
        testID="mocked-svg-icon"
        width={24}
      />
    </View>
  </ViewManagerAdapter_ExpoGlassEffect>
</View>
`;
